// fus-ext/test-target/manifest.js
"use strict";
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.global = exports.clearImmediate = exports.setImmediate = exports.dotDot_573300145710716007 = exports.fePromiseStatic = exports.feUint8ArrayStatic = exports.feUint8Array = exports.feConsole = exports.feDate = exports.feJson = exports.feMath = exports.feArray = exports.feObject = exports.feString = exports.feNumberStatic = exports.feNumber = exports.cmath = exports.web = exports.sys = exports.AsyncQueue = exports.Point = exports.Observer = exports.IntervalTimer = exports.Timer = exports.eventField = exports.spread = exports.assert = exports.compose = exports.fail = exports.enum = exports.break = exports.repeat = exports.loop = undefined;

var _main = require("../target/main");

var var_573300145710716007_2;var var_573300145710716007_4;var global, var_573300145710716007_3, var_573300145710716007_5;exports.loop = _main.loop;
exports.repeat = _main.repeat;
exports.break = _main.break;
exports.enum = _main.enum;
exports.fail = _main.fail;
exports.compose = _main.compose;
exports.assert = _main.assert;
exports.spread = _main.spread;
exports.eventField = _main.eventField;
exports.setImmediate = var_573300145710716007_2 = (var_573300145710716007_3 = _main.sys.global.setImmediate) === undefined ? _main.setImmediate : var_573300145710716007_3;exports.clearImmediate = var_573300145710716007_4 = (var_573300145710716007_5 = _main.sys.global.clearImmediate) === undefined ? _main.clearImmediate : var_573300145710716007_5;exports.Timer = _main.Timer;
exports.IntervalTimer = _main.IntervalTimer;
exports.Observer = _main.Observer;
exports.Point = _main.Point;
exports.AsyncQueue = _main.AsyncQueue;
exports.sys = _main.sys;
exports.web = _main.web;
exports.cmath = _main.cmath;
exports.feNumber = _main.feNumber;
exports.feNumberStatic = _main.feNumberStatic;
exports.feString = _main.feString;
exports.feObject = _main.feObject;
exports.feArray = _main.feArray;
exports.feMath = _main.feMath;
exports.feJson = _main.feJson;
exports.feDate = _main.feDate;
exports.feConsole = _main.feConsole;
exports.feUint8Array = _main.feUint8Array;
exports.feUint8ArrayStatic = _main.feUint8ArrayStatic;
exports.fePromiseStatic = _main.fePromiseStatic;
exports.dotDot_573300145710716007 = _main.dotDot_573300145710716007;
exports.global = global = _main.sys.global;exports.setImmediate = var_573300145710716007_2;
exports.clearImmediate = var_573300145710716007_4;
exports.global = global;
